### PR TITLE
Support for CSV and other dialects in tabular

### PR DESCRIPTION
**Pull request by vavrusa**
## Fixes
- separator can be changed a/o sniffed from the dataset (preview and table view)
- fixed hardcoded separator in the dataproviders (visualization)
- skipped rows are not displayed in the preview
- sniffing of dataset column names from the header
- sniffer for CSV datasets containing a header
## Enhances
- `stats/gsummary` works with [RPy2](http://rpy.sourceforge.net/rpy2/doc-dev/html/overview.html) classic mode, as the original RPy is [pretty much dead](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=705847)
- it also supports arbitrary CSV dialects

TL;DR works well with all CSV dialects now and is able to sniff some of them
## How to test

I have attached some example files at https://gist.github.com/vavrusa/f8bb8d059c6c96dc7585
to test it, the one with the first line as a header should be sniffed correctly when uploading, the rest should be correctly displayed when saved as `tabular`. Visualization for all should be correct.

### CODE REVIEW

File: config/datatypes_conf.xml.sample [hunk 1]
```diff
@@ -51,7 +51,7 @@
       <converter file="bed_gff_or_vcf_to_bigwig_converter.xml" target_datatype="bigwig"/>
     </datatype>
     <!-- MSI added Datatypes -->
-    <datatype extension="csv" type="galaxy.datatypes.tabular:Tabular" subclass="True" display_in_upload="true" /> <!-- FIXME: csv is 'tabular'ized data, but not 'tab-delimited'; the class used here is intended for 'tab-delimited' -->
+    <datatype extension="csv" type="galaxy.datatypes.tabular:CSV" subclass="True" display_in_upload="true" />
```
Comment (martenson) [inline config/datatypes_conf.xml.sample:54]: This requires discussion because as it is still a subtype of tabular tools will expect from this type to actually be separated by tabs and probably won't work otherwise. It might be viable to leave the old Tabular datatype be and subclass galaxy.datatypes.text into a new datatype instead. 
@blankenberg has more insights, please comment.

Comment (vavrusa) [inline config/datatypes_conf.xml.sample:54]: I think all (if not most) tabular tools have been converted to work with variable separator, my idea is that the same tools should work with **any** tabular dialect. Maybe more work has to be done on this, but I have yet to see a project where CSV, TSV, ... are treated as different types, e.g. _shell, awk, shell, R, ..._ all work the same on the same dataset with any specified separator. I think the sources should use more of the [csv](https://docs.python.org/2/library/csv.html) package that's in Python since 2.3 instead of handwritten parsing and all should be fine.

**Comment by carlfeberhard** [comment]
Hey, vavrusa: good stuff overall I think! Thank you. I'm looking forward to further discussion on the inheritance issue.

A few small things:
- When you said, "fixed hardcoded separator in the dataproviders (visualization)", it's the opposite - you've actually made it so one can't pass in a delimiter via the query in the dataproviders (your delimiter is always used).
- "skipped rows are not displayed in the preview" - I'm not sure we do it that way for other previews/peeks, but someone else may know more.
- You should catch specific errors if you're going to use a 'pass' in a try/except.

nits: 
- You've got magic numbers in the new code. A descriptive class var might be good there.

> I think the sources should use more of the csv package that's in Python since 2.3 instead of handwritten parsing and all should be fine.

I'm not sure that's 100% true. At least in the case of the dataproviders, there's a reason to not rely on csv for the parsing IIRC. That may not apply to the larger question of sniffing and class inheritance, tho.

**Comment by vavrusa** [comment]
Hey, no worries! I'd like to make Galaxy better not worse, so don't go easy on me! 
- So the queries can specify an arbitrary delimiter? I'm not sure how the dataproviders interact with the front-end, but I suspect the front-end just asks for a piece of data and the dataprovider is going to provide it in some universal format (list or map), so the actual file type is abstracted out. At least I couldn't find any such option in the visualizations. So what do you think is a best course of action? Check if the delimiter is set in the query and set to sniffed one if it isn't?
- I'm not sure either, but if there's a functionality to skip rows, then they should be skipped everywhere not just in metadata discovery?
- Right, you mean the try/except for sniffing separator? I can add specific errors, but I'm not sure if it's interesting - the only interesting outcome is "Can you guess the separator?" "yes/no" 899cc95
- Alright, 899cc95

So I'm new to Galaxy and I'd like to hear your opinion on what should the `tabular` type represent.
From my perspective it would be very convenient if it represented any tabular file, since I have to deal with CSVs, TSVs, semicolons, spaces ... and I'd like to have statistics and visualizations regardless of the type and avoid reimplementing dataproviders for each class. I'm writing support for molecular structure datatypes as well, so I'd like to do it properly.

File: lib/galaxy/datatypes/tabular.py [hunk 2]
```diff
@@ -285,7 +322,6 @@ def display_data(self, trans, dataset, preview=False, filename=None, to_ext=None
             #Fancy tabular display is only suitable for datasets without an incredibly large number of columns.
             #We should add a new datatype 'matrix', with its own draw method, suitable for this kind of data.
             #For now, default to the old behavior, ugly as it is.  Remove this after adding 'matrix'.
-            max_peek_size = 1000000 # 1 MB
             if os.stat( dataset.file_name ).st_size < max_peek_size:
```
Comment (martenson) [inline lib/galaxy/datatypes/tabular.py]: self.max_peek_size

**Comment by blankenberg** [comment]
We can't change Tabular to allow anything other than tab-delimited data without breaking lots of tools that are outside of our control. A tool that accepts 'tabular' datasets is expected to be given a file that is delimited with tabs and nothing else; if 'tabular' datasets can now instead be e.g. comma separated, then tools that are using the 'tabular' datatype will get data that they are not able to handle, defeating the purpose of it even restricting its input to 'tabular'; in other words, we just turned 'tabular' into 'text'.

from tabular.py:

```
class Tabular( data.Text ):
    """Tab delimited data"""
```

Now, we could do something like:

```
class TabularizedData( data.Text ):
    """Various Tabularized data"""
       ......<this new datatype as defined here>...

class Tabular( TabularizedData ):
    """Tab delimited data"""
   ...current behavior, delimiter is always restricted to '\t'
```

but I am not sure if this gets you closer to where you are trying to go, as all existing datatypes that currently subclass from Tabular would still need to, as they expect tab-delimited files.

And it is not possible to 'fix all the tools' to allow them to work with this not-always-tab-delimited 'tabular' type -- we don't control, nor know about all of the tools using the tabular datatype.

**Comment by vavrusa** [comment]
I wonder how many of these tools will be suddenly broken since nobody uses CSV files in the first place.
Then if people started using CSVs, they would need fix the tools first.
It's just a pity because it could have been a single class if it was documented better, there are `TODO`s in the code suggesting that it was the original intention. Very well, if you feel like tabular type is more like _taboolar_ then I'm going to reimplement a new class with proper abstraction.

**Comment by martenson** [comment]
There are cases where people upload csv and then use `Convert delimiters to TAB` tool to make it tabular (in current Galaxy meaning). If we just switched they would be able to select the .csv as direct input and the following tool fails.
We can kick the idea around for a bit but I don't see an easy way out of here yet.

@jmchilton might be interested, also ping @jj-umn the original author

**Comment by jmchilton** [comment]
- Having a more generic tabular base class like @blankenberg describd does help I think. He is right that it doesn't help subclasses - but most of the subclasses are actual expected to be tab-deliniated by default (e.g. [bed](https://genome.ucsc.edu/FAQ/FAQformat.html#format1)). Having a correct super class and converting the standard tools to allow consumption of these would help applications that would benefit from other kinds of tabular datatypes.
- Can we all agree that having a `csv` type that must be tab delinated by default is a bug - the c in csv doesn't stand for `TAB`?

There may be something I am missing - but my initial thought is that I would be :+1: on [extracting a generic superclass](https://sourcemaking.com/refactoring/extract-superclass) of `Tabular`, having CSV extend that, and starting to work on converting the basic tabular operations to allow the more generic tabular datatype if they can. I have worked with tools in the past that dealt with comma-separated values and I just had to convert them to tabs before the job was finished - this is fine but my users wanted the raw outputs and were annoyed I had to convert them.

**Comment by vavrusa** [comment]
> There may be something I am missing - but my initial thought is that I would be :+1: on extracting a
> generic superclass of Tabular, having CSV extend that, and starting to work on converting the basic
> tabular operations to allow the more generic tabular datatype if they can.

OK then, sounds sane to me, I'll redo the PR.

> I have worked with tools in the past that dealt with comma-separated values and I just had to convert
> them to tabs before the job was finished - this is fine but my users wanted the raw outputs and were 
> annoyed I had to convert them.

This, I have a bunch of different formats and it's awkward to always convert stuff in between because tool A wants this and tool B that. It's all just a different notation for table data anyway.
